tutorials-v5/time-evolution/023_v5_paper-mesolve.md

```diff
@@ -0,0 +1,565 @@
+c_ops_me_RWA = [np.sqrt(gamma) * sigmam()]
+```
+
+```python
+driv_RWA_res = mesolve(H_RWA, psi0, tlist, c_ops=c_ops_me_RWA, e_ops=e_ops)
+```
+
+### With `brmesolve`
+
+```python
+# Bose einstein distribution
+def nth(w):
+    if temp > 0:
+        return 1 / (np.exp(w / temp) - 1)
+    else:
+        return 0
+
+# Power spectrum
+def power_spectrum(w):
+    if w > 0:
+        return gamma * (nth(w) + 1)
+    elif w == 0:
+        return 0
+    else:
+        return gamma * nth(-w)
+```
+
+```python
+a_ops = [[sigmax(), power_spectrum]]
+```
+
+```python
+driv_br_res = brmesolve(H, psi0, tlist, a_ops, sec_cutoff=-1, e_ops=e_ops)
+```
+
+### With `HEOMSolver`
+
+```python
+max_depth = 4  # number of hierarchy levels
+
+wsamp = 2 * np.pi
+w0 = 5 * 2 * np.pi
+gamma_heom = 1.9 * w0
+
+lambd = np.sqrt(
+    0.5 * gamma / (gamma_heom * wsamp)
+    * ((w0**2 - wsamp**2) ** 2 + (gamma_heom**2) * ((wsamp) ** 2))
+)
+```
+
+```python
+# Create Environment
+bath = UnderDampedEnvironment(lam=lambd, w0=w0, gamma=gamma_heom, T=1e-10)
+fit_times = np.linspace(0, 5, 1000)  # range for correlation function fit
+
+# Fit correlation function with exponentials
+exp_bath, fit_info = bath.approx_by_cf_fit(fit_times, Ni_max=1, Nr_max=2, target_rsme=None)
+print(fit_info["summary"])
+```
+
+```python
+HEOM_corr_fit = HEOMSolver(QobjEvo(H), (exp_bath, sigmax()), max_depth=max_depth,
+                    options={'nsteps': 15000, 'rtol': 1e-12, 'atol': 1e-12})
+results_corr_fit = HEOM_corr_fit.run(psi0 * psi0.dag(), tlist, e_ops=e_ops)
+```
+
+### Comparison of Solver Results
+
+```python
+plt.figure()
+
+plt.plot(tlist, driv_res.expect[0], "-", label="mesolve (time-dep)")
+plt.plot(tlist, driv_RWA_res.expect[0], "-.", label="mesolve (rwa)")
+plt.plot(tlist, np.real(results_corr_fit.expect[0]), '--', label=r'heomsolve')
+plt.plot(tlist, driv_br_res.expect[0], ":", linewidth=3, label="brmesolve")
+
+plt.xlabel(r"$t\, /\, \Delta^{-1}$")
+plt.ylabel(r"$\langle \sigma_z \rangle$")
+plt.legend()
+plt.show()
+```
+
+### Adiabatic Energy Switching
+
+To show where usage of these local-basis collapse operators can lead to errors, we look at a single qubit system.
+This time however, the energies of the qubit are adiabatically switched between positive and negative values.
+
+$H = \dfrac{\Delta}{2} \sin{(\omega_d t)} \sigma_z$.
+
+If the drive is slow enough, we expect the bath to respond to this change and transitions from higher to lower energy levels should be induced.
+
+```python
+# Hamiltonian
+omega_d = 0.05 * Delta  # drive frequency
+A = Delta  # drive amplitude
+H_adi = [[A / 2.0 * sigmaz(), f]]
+
+# Bath parameters
+gamma = 0.05 * Delta / (2 * np.pi)
+
+# Simulation parameters
+T = 2 * np.pi / omega_d  # period length
+tlist = np.linspace(0, 2 * T, 400)
+```
+
+```python
+# Simple mesolve
+c_ops_me = [np.sqrt(gamma) * sigmam()]
+adi_me_res = mesolve(H_adi, psi0, tlist, c_ops=c_ops_me, e_ops=e_ops)
+```
+
+```python
+# HEOM
```

Comment:
Since the parameter `gamma` is slightly changed we need to refit the heom bath here, i.e., have again:

```
max_depth = 4  # number of hierarchy levels

wsamp = 2 * np.pi
w0 = 5 * 2 * np.pi
gamma_heom = 1.9 * w0

lambd = np.sqrt(
    0.5 * gamma / (gamma_heom * wsamp)
    * ((w0**2 - wsamp**2) ** 2 + (gamma_heom**2) * ((wsamp) ** 2))
)
# Create Environment
bath = UnderDampedEnvironment(lam=lambd, w0=w0, gamma=gamma_heom, T=1e-10)
fit_times = np.linspace(0, 5, 1000)  # range for correlation function fit

# Fit correlation function with exponentials
exp_bath, fit_info = bath.approx_by_cf_fit(fit_times, Ni_max=1, Nr_max=2, target_rsme=None)
print(fit_info["summary"])
```